enhancements/kube-apiserver/stale-clusteroperators-detection.md

```diff
@@ -0,0 +1,108 @@
+---
+title: detection-and-reporting-of-stale-clusteroperators
+authors:
+  - "@mfojtik"
+reviewers:
+  - "@deads2k"
+approvers:
+  - "@deads2k"
+creation-date: 2022-01-31
+last-updated: 2022-01-31
+---
+
+# Detection and reporting of stale cluster operators
+
+## Summary
+
+In OpenShift 4, every active operator included in the core payload (cluster operator) updates its conditions in "
+clusteroperator" resource. Conditions determine whether specific core components are Degraded or Available or if they
+are making progress or if their state causes the cluster to be not upgradeable. These conditions are essential for
+cluster administrators to determine the health of their clusters, and they are usually the first thing Red Hat support
+is going to look at in case there is a problem with the cluster.
+
+Unfortunately, in some edge cases, these conditions might not reflect the current state of the cluster. For example,
+since they depend on the running operator, these conditions might become stale and outdated if the operator is not
+running or cannot contact the API server. And while the API server might still work, the result
```

Comment:
ClusterOperator is primarily an API for operators to report to the cluster-version operator.  The CVO will expose concerning issues like `Available=False` or `Degraded=True` to admins in ClusterVersion.  But the CVO also manages the operator Deployments, so in the case where the operator is actually not running, the deployment controller should be reporting some sad condition, and the CVO should be complaining about that sad deployment in ClusterVersion's `Failing`.  Admins that reach around the CVO and inspect ClusterOperator directly may not be aware of the Deployment context, but :shrug: that's what you get when you imperfectly break through an abstraction layer :wink:.  And there are some cases today, when the CVO is stuck mid-update on some earlier manifest, where even the CVO stops caring about Deployments and ClusterOperators that are later in the update order (at least as far as `Failing` reporting goes).  But I think all of that either works today, or could work tomorrow if we are concerned about it.

The remaining issue, which this `message`-poking approach could help with, is where the operator deployment is up and happy, but the loop in charge of reporting status conditions has wedged.  But not clear to me that `message`-poking is a better way to detect that vs. a liveness probe or similar.  And the benefit of using a liveness probe is that if the operator locks up, Kube will cycle the container for you to help recover, instead of just complaining about the lockup.

Then there's also a possibility that the boring liveness service keeps working, but the more complicated status sync loop locks up.  Maybe `message`-poking is the best way at that?  But by this point, we're down in what seems like a pretty tiny corner, and we may be fine just living with the risk of the occasional partially-locked-operator bug.